src/main/java/org/broadinstitute/hellbender/tools/copynumber/ModelSegments.java

```diff
@@ -0,0 +1,570 @@
+            shortName = NUM_SAMPLES_ALLELE_FRACTION_SHORT_NAME,
+            optional = true,
+            minValue = 1
+    )
+    private int numSamplesAlleleFraction = 100;
+
+    @Argument(
+            doc = "Number of burn-in samples to discard for allele-fraction model.",
+            fullName = NUM_BURN_IN_ALLELE_FRACTION_LONG_NAME,
+            shortName = NUM_BURN_IN_ALLELE_FRACTION_SHORT_NAME,
+            optional = true,
+            minValue = 0
+    )
+    private int numBurnInAlleleFraction = 50;
+
+    @Argument(
+            doc = "Number of 10% equal-tailed credible-interval widths to use for copy-ratio segmentation smoothing.",
+            fullName = SMOOTHING_CREDIBLE_INTERVAL_THRESHOLD_COPY_RATIO_LONG_NAME,
+            shortName = SMOOTHING_CREDIBLE_INTERVAL_THRESHOLD_COPY_RATIO_SHORT_NAME,
+            optional = true,
+            minValue = 0.
+    )
+    private double smoothingCredibleIntervalThresholdCopyRatio = 2.;
+
+    @Argument(
+            doc = "Number of 10% equal-tailed credible-interval widths to use for allele-fraction segmentation smoothing.",
+            fullName = SMOOTHING_CREDIBLE_INTERVAL_THRESHOLD_ALLELE_FRACTION_LONG_NAME,
+            shortName = SMOOTHING_CREDIBLE_INTERVAL_THRESHOLD_ALLELE_FRACTION_SHORT_NAME,
+            optional = true,
+            minValue = 0.
+    )
+    private double smoothingCredibleIntervalThresholdAlleleFraction = 2.;
+
+    @Argument(
+            doc = "Maximum number of iterations allowed for segmentation smoothing.",
+            fullName = MAX_NUM_SMOOTHING_ITERATIONS_LONG_NAME,
+            shortName = MAX_NUM_SMOOTHING_ITERATIONS_SHORT_NAME,
+            optional = true,
+            minValue = 0
+    )
+    private int maxNumSmoothingIterations = 25;
+
+    @Argument(
+            doc = "Number of segmentation-smoothing iterations per MCMC model refit. " +
+                    "(Increasing this will decrease runtime, but the final number of segments may be higher. " +
+                    "Setting this to 0 will completely disable model refitting between iterations.)",
+            fullName = NUM_SMOOTHING_ITERATIONS_PER_FIT_LONG_NAME,
+            shortName = NUM_SMOOTHING_ITERATIONS_PER_FIT_SHORT_NAME,
+            optional = true,
+            minValue = 0
+    )
+    private int numSmoothingIterationsPerFit = 0;
+
+    //initialize data variables, some of which may be optional
+    private CopyRatioCollection denoisedCopyRatios = null;
+    private AllelicCountCollection hetAllelicCounts = null;
+
+    @Override
+    protected Object doWork() {
+        validateArguments();
+
+        //perform one-dimensional or multidimensional segmentation as appropriate and write to file
+        //(for use by CallCopyRatioSegments, if copy ratios are available)
+        final MultidimensionalSegmentCollection multidimensionalSegments;
+        if (inputDenoisedCopyRatiosFile != null && inputAllelicCountsFile == null) {
+            readDenoisedCopyRatios();
```

Comment:
I do prefer more functional code and I feel like this and other methods calls below (such as `readDenoisedCopyRatios` and `readAndFilterAllelicCounts`) should not work through side effects

Reply:
I agree, the control flow could probably be improved here and in the gCNV CLIs.  Filed #3951.